src/main/java/build/buildfarm/server/OperationQueueService.java

```diff
@@ -88,12 +88,18 @@ public void put(
       return;
     }
 
-    boolean ok = instance.putOperation(operation);
-    Code code = ok ? Code.OK : Code.UNAVAILABLE;
-    responseObserver.onNext(com.google.rpc.Status.newBuilder()
-        .setCode(code.getNumber())
-        .build());
-    responseObserver.onCompleted();
+    try {
+      boolean ok = instance.putOperation(operation);
+      Code code = ok ? Code.OK : Code.UNAVAILABLE;
+      responseObserver.onNext(com.google.rpc.Status.newBuilder()
+          .setCode(code.getNumber())
+          .build());
+      responseObserver.onCompleted();
+    } catch (IllegalStateException e) {
+      responseObserver.onError(Status.FAILED_PRECONDITION
```

Comment:
We also want to send a PreconditionFailure message in the status error_detail field, with a Violation of type MISSING and the right digest where appropriate (according to the API spec), so that the client can reupload+retry that.
I think the simplest way of doing that would be to raise and handle a CacheNotFoundException.

Reply:
Ah, yeah.  I was going to switch to making a PreconditionViolationException or some such to do those aggregates and lump as many issues as possible into one response